In [1]:
import pandas as pd
import pickle
pd.set_option('display.max_columns', None)



### Reading and drop na columns
file_names = ['dataframes_list_season_-2017.pkl',
              'dataframes_list_season_-2018.pkl', 'dataframes_list_season_-2019.pkl',
              'dataframes_list_season_-2020.pkl', 'dataframes_list_season_-2021.pkl',
              'dataframes_list_season_-2022.pkl','dataframes_list_season_-2023.pkl']

all_cleaned_dataframes = []

for file_name in file_names:
    try:
        with open("Data/"+file_name, 'rb') as file:
            # Load the dataframe from pickle file
            dataframe = pickle.load(file)

            # Filter out columns that start with 'Unnamed:'
            dataframe = dataframe.loc[:, ~dataframe.columns.str.startswith('Unnamed:')]

            # Drop all columns that are entirely NA
            dataframe = dataframe.dropna(axis=1, how='all')

            # Add the cleaned dataframe to the list
            all_cleaned_dataframes.append(dataframe)
            
            ## print(f"Processed {file_name}, columns: {dataframe.columns}")

    except Exception as e:
        print(f"Error processing {file_name}: {e}")

# Concatenate all dataframes into one
df = pd.concat(all_cleaned_dataframes, ignore_index=True)



### delete some more columns
columns_to_delete = ['OT', 'OT_opp', '2OT', '3OT', '2OT_opp', '3OT_opp', '4OT', '4OT_opp',
                    'mp_total_opp','bpm_max','bpm_max_opp']

# Drop the specified columns from the dataframe
df.drop(columns=columns_to_delete, inplace=True)

#### rename columns
df.rename(columns = {'mp_total':'mp'}, inplace=True)

print("data shape:", df.shape)

columns_format = list(df.columns)

df.head(2)

data shape: (17760, 149)


mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
0  240.0      32.0       87.0      0.368       9.0       27.0      0.333   
1  240.0      45.0       94.0      0.479      13.0       35.0      0.371   

   ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  ast_total  \
0      15.0       20.0      0.750       13.0       29.0       42.0       17.0   
1      14.0       19.0      0.737       11.0       40.0       51.0       31.0   

   stl_total  blk_total  tov_total  pf_total  pts_total  ts%_total  \
0        6.0        6.0       18.0      22.0       88.0      0.459   
1       12.0        5.0       14.0      22.0      117.0      0.572   

   efg%_total  3par_total  ftr_total  orb%_total  drb%_total  trb%_total  \
0       0.420       0.310      0.230        24.5        72.5        45.2   
1       0.548       0.372      0.202        27.5        75.5        54.8   

   ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  ortg_total  \
0        53.1         6.0        10.2        15.8       100.0        88.1   
1        68.9        12.0         8.3        12.0       100.0       117.1   

   drtg_total  fg_max  fga_max  fg%_max  3p_max  3pa_max  3p%_max  ft_max  \
0       117.1     8.0     18.0    0.667     3.0      5.0      1.0     5.0   
1        88.1    12.0     22.0    1.000     4.0      8.0      1.0     9.0   

   fta_max  ft%_max  orb_max  drb_max  trb_max  ast_max  stl_max  blk_max  \
0      6.0      1.0      4.0      5.0      7.0      5.0      2.0      2.0   
1     12.0      1.0      3.0     10.0     12.0     14.0      3.0      1.0   

   tov_max  pf_max  pts_max  +/-_max  ts%_max  efg%_max  3par_max  ftr_max  \
0      4.0     5.0     19.0     -1.0    0.833     0.833     1.000      2.0   
1      4.0     4.0     29.0     26.0    1.000     1.000     0.833      1.0   

   orb%_max  drb%_max  trb%_max  ast%_max  stl%_max  blk%_max  tov%_max  \
0      11.0      31.2      16.8      45.5       5.8      10.6      50.0   
1      17.0      36.3      24.8      65.5       6.4      14.4      66.7   

   usg%_max  ortg_max  drtg_max team   1   2   3   4 total  home  \
0      36.8     153.0     125.0  NYK  18  27  19  24    88     0   
1      36.8     200.0     101.0  CLE  28  20  34  35   117     1   

   fg_total_opp  fga_total_opp  fg%_total_opp  3p_total_opp  3pa_total_opp  \
0          45.0           94.0          0.479          13.0           35.0   
1          32.0           87.0          0.368           9.0           27.0   

   3p%_total_opp  ft_total_opp  fta_total_opp  ft%_total_opp  orb_total_opp  \
0          0.371          14.0           19.0          0.737           11.0   
1          0.333          15.0           20.0          0.750           13.0   

   drb_total_opp  trb_total_opp  ast_total_opp  stl_total_opp  blk_total_opp  \
0           40.0           51.0           31.0           12.0            5.0   
1           29.0           42.0           17.0            6.0            6.0   

   tov_total_opp  pf_total_opp  pts_total_opp  ts%_total_opp  efg%_total_opp  \
0           14.0          22.0          117.0          0.572           0.548   
1           18.0          22.0           88.0          0.459           0.420   

   3par_total_opp  ftr_total_opp  orb%_total_opp  drb%_total_opp  \
0           0.372          0.202            27.5            75.5   
1           0.310          0.230            24.5            72.5   

   trb%_total_opp  ast%_total_opp  stl%_total_opp  blk%_total_opp  \
0            54.8            68.9            12.0             8.3   
1            45.2            53.1             6.0            10.2   

   tov%_total_opp  usg%_total_opp  ortg_total_opp  drtg_total_opp  fg_max_opp  \
0            12.0           100.0           117.1            88.1        12.0   
1            15.8           100.0            88.1           117.1         8.0   

   fga_max_opp  fg%_max_opp  3p_max_opp  3pa_max_opp  3p%_max_opp  ft_max_opp  \
0         22.0        1.000         4.

In [2]:
df['+/-_max'].isnull().sum()

15

In [3]:
df['+/-_max_opp'].isnull().sum()

15

In [4]:
##### Abbrivate the Team names
team_df = pd.read_excel('Team_full-forms.xlsx')
team_df['team'] = team_df['team'].str.strip()
team_df['team1'] = team_df['team1'].str.strip()


##### Merge and delete the columns
df = pd.merge(team_df, df, on = ['team'], how='inner')
del df['team']
df.rename(columns = {'team1':'team'}, inplace=True)

team_df.rename(columns = {'team':'team_opp'}, inplace=True)
df = pd.merge(team_df, df, on = ['team_opp'], how='inner')
del df['team_opp']
df.rename(columns = {'team1':'team_opp'}, inplace=True)

print("data shape:", df.shape)

df = df[columns_format]

## ordering with date
df['date'] = pd.to_datetime(df['date']).dt.date
df = df.sort_values(by = ['date'], ascending=True).reset_index(drop=True)

df

data shape: (17760, 149)


mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
0      240.0      32.0       87.0      0.368       9.0       27.0      0.333   
1      240.0      47.0       98.0      0.480      12.0       24.0      0.500   
2      240.0      40.0       85.0      0.471       7.0       33.0      0.212   
3      240.0      40.0       82.0      0.488       8.0       24.0      0.333   
4      240.0      45.0       94.0      0.479      13.0       35.0      0.371   
...      ...       ...        ...        ...       ...        ...        ...   
17755  240.0      41.0       80.0      0.513       5.0       18.0      0.278   
17756  240.0      35.0       78.0      0.449       8.0       25.0      0.320   
17757  240.0      39.0       79.0      0.494      14.0       28.0      0.500   
17758  240.0      38.0       84.0      0.452       5.0       28.0      0.179   
17759  240.0      33.0       96.0      0.344       9.0       35.0      0.257   

       ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  \
0          15.0       20.0      0.750       13.0       29.0       42.0   
1          23.0       26.0      0.885       21.0       34.0       55.0   
2          13.0       18.0      0.722        8.0       27.0       35.0   
3          16.0       16.0      1.000        6.0       25.0       31.0   
4          14.0       19.0      0.737       11.0       40.0       51.0   
...         ...        ...        ...        ...        ...        ...   
17755      22.0       27.0      0.815       13.0       45.0       58.0   
17756      17.0       20.0      0.850        8.0       29.0       37.0   
17757      16.0       21.0      0.762        5.0       29.0       34.0   
17758      13.0       23.0      0.565       11.0       46.0       57.0   
17759      14.0       16.0      0.875       11.0       33.0       44.0   

       ast_total  stl_total  blk_total  tov_total  pf_total  pts_total  \
0           17.0        6.0        6.0       18.0      22.0       88.0   
1           25.0       13.0        3.0       13.0      19.0      129.0   
2           24.0       11.0        6.0       16.0      19.0      100.0   
3           19.0        9.0        5.0       11.0      19.0      104.0   
4           31.0       12.0        5.0       14.0      22.0      117.0   
...          ...        ...        ...        ...       ...        ...   
17755       28.0        3.0        5.0       13.0      18.0      109.0   
17756       23.0        2.0        3.0       14.0      19.0       95.0   
17757       26.0       11.0        7.0        6.0      18.0      108.0   
17758       21.0        6.0        7.0       14.0      13.0       94.0   
17759       18.0        9.0        7.0        8.0      21.0       89.0   

       ts%_total  efg%_total  3par_total  ftr_total  orb%_total  drb%_total  \
0          0.459       0.420       0.310      0.230        24.5        72.5   
1          0.589       0.541       0.245      0.265        43.8        81.0   
2          0.538       0.512       0.388      0.212        19.0        56.3   
3          0.584       0.537       0.293      0.195        17.1        83.3   
4          0.572       0.548       0.372      0.202        27.5        75.5   
...          ...         ...         ...        ...         ...         ...   
17755      0.593       0.544       0.225      0.338        36.1        81.8   
17756      0.547       0.500       0.321      0.256        21.6        85.3   
17757      0.612       0.582       0.354      0.266        14.7        78.4   
17758      0.499       0.482       0.333      0.274        25.0        80.7   
17759      0.432       0.391       0.365      0.167        19.3        75.0   

       trb%_total  ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  \
0            45.2        53.1         6.0        10.2        15.8       100.0   
1            61.1        53.2        13.2         5.8        10.6       100.0   
2            38.9        60.0        11.2         8.1        14.7       100.0  

1. We will create a target column (needs to be created on team level) -> represents the next game outcome. (Won column indicates the current match, target column indicates next match)

2. Replace the Null values in Target column with “2”, False (Loss) = 0, True (Won) = 1.


In [5]:
def add_target(team):
    team['target'] = team['won'].shift(-1)
    return team

df = df.groupby("team", group_keys=False).apply(add_target)

## Preprocessing Target column (Null = 2, True = 1, False = 0)

df['target'][pd.isnull(df['target'])] = 2
df['target'] = df['target'].astype(int, errors='ignore')

C:\Users\JETENDRA\AppData\Local\Temp\ipykernel_50684\1840707839.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['target'][pd.isnull(df['target'])] = 2


In [6]:
### checking the data is balance / Imbalanced

df['won'].value_counts()

True     8880
False    8880
Name: won, dtype: int64

In [7]:
df['target'].value_counts()

1    8865
0    8865
2      30
Name: target, dtype: int64

Checking Null values and dropping columns and rows

In [8]:
### Checking null values

null_columns = df.isnull().sum()
null_columns[null_columns > 0]

+/-_max           15
+/-_max_opp       15
index_opp      10238
dtype: int64

In [9]:
### delete some more columns
more_columns_to_delete = ['index_opp']

# Drop the specified columns from the dataframe
df.drop(columns=more_columns_to_delete, inplace=True)

## as we have only 1 null row (match) we will drop it
df = df.dropna()

null_columns = df.isnull().sum()
null_columns[null_columns > 0]

Series([], dtype: int64)

In [10]:
## re-ordering on date

## ordering with date
df['date'] = pd.to_datetime(df['date']).dt.date
df = df.sort_values(by = ['date'], ascending=True).reset_index(drop=True)

print("data shape:", df.shape)

df.head()

data shape: (17744, 149)


mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
0  240.0      32.0       87.0      0.368       9.0       27.0      0.333   
1  240.0      47.0       98.0      0.480      12.0       24.0      0.500   
2  240.0      40.0       85.0      0.471       7.0       33.0      0.212   
3  240.0      40.0       82.0      0.488       8.0       24.0      0.333   
4  240.0      45.0       94.0      0.479      13.0       35.0      0.371   

   ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  ast_total  \
0      15.0       20.0      0.750       13.0       29.0       42.0       17.0   
1      23.0       26.0      0.885       21.0       34.0       55.0       25.0   
2      13.0       18.0      0.722        8.0       27.0       35.0       24.0   
3      16.0       16.0      1.000        6.0       25.0       31.0       19.0   
4      14.0       19.0      0.737       11.0       40.0       51.0       31.0   

   stl_total  blk_total  tov_total  pf_total  pts_total  ts%_total  \
0        6.0        6.0       18.0      22.0       88.0      0.459   
1       13.0        3.0       13.0      19.0      129.0      0.589   
2       11.0        6.0       16.0      19.0      100.0      0.538   
3        9.0        5.0       11.0      19.0      104.0      0.584   
4       12.0        5.0       14.0      22.0      117.0      0.572   

   efg%_total  3par_total  ftr_total  orb%_total  drb%_total  trb%_total  \
0       0.420       0.310      0.230        24.5        72.5        45.2   
1       0.541       0.245      0.265        43.8        81.0        61.1   
2       0.512       0.388      0.212        19.0        56.3        38.9   
3       0.537       0.293      0.195        17.1        83.3        47.7   
4       0.548       0.372      0.202        27.5        75.5        54.8   

   ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  ortg_total  \
0        53.1         6.0        10.2        15.8       100.0        88.1   
1        53.2        13.2         5.8        10.6       100.0       131.3   
2        60.0        11.2         8.1        14.7       100.0       101.8   
3        47.5         9.9         8.9        11.0       100.0       114.9   
4        68.9        12.0         8.3        12.0       100.0       117.1   

   drtg_total  fg_max  fga_max  fg%_max  3p_max  3pa_max  3p%_max  ft_max  \
0       117.1     8.0     18.0    0.667     3.0      5.0     1.00     5.0   
1       101.8    10.0     21.0    1.000     3.0      6.0     1.00    15.0   
2       131.3    11.0     18.0    1.000     3.0     10.0     0.40     5.0   
3       124.8    12.0     18.0    1.000     3.0      7.0     0.75     5.0   
4        88.1    12.0     22.0    1.000     4.0      8.0     1.00     9.0   

   fta_max  ft%_max  orb_max  drb_max  trb_max  ast_max  stl_max  blk_max  \
0      6.0      1.0      4.0      5.0      7.0      5.0      2.0      2.0   
1     15.0      1.0      8.0      6.0     14.0      5.0      5.0      2.0   
2      6.0      1.0      5.0     10.0     12.0      6.0      5.0      2.0   
3      5.0      1.0      4.0     10.0     14.0      6.0      2.0      2.0   
4     12.0      1.0      3.0     10.0     12.0     14.0      3.0      1.0   

   tov_max  pf_max  pts_max  +/-_max  ts%_max  efg%_max  3par_max  ftr_max  \
0      4.0     5.0     19.0     -1.0    0.833     0.833     1.000    2.000   
1      2.0     4.0     35.0     35.0    1.500     1.500     1.000    0.714   
2      4.0     4.0     27.0     -3.0    1.000     1.000     0.667    0.500   
3      3.0     5.0     29.0      7.0    1.000     1.000     0.412    0.571   
4      4.0     4.0     29.0     26.0    1.000     1.000     0.833    1.000   

   orb%_max  drb%_max  trb%_max  ast%_max  stl%_max  blk%_max  tov%_max  \
0      11.0      31.2      16.8      45.5       5.8      10.6      50.0   
1      26.9      31.0      28.7      55.5      10.9      11.1     100.0   
2      17.5      27.3      19.6      30.8       7.5       7.2      75.0   
3      16.5      39.1     

1. we are using TimeSeriesSplit as we need to split as dates only
2. reason: we don't want to split the old and new data in mix, we want only historic data to predict

In [11]:
from sklearn.linear_model import RidgeClassifier
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.model_selection import TimeSeriesSplit


rr = RidgeClassifier(alpha=1)

split = TimeSeriesSplit(n_splits=3)

sfs = SequentialFeatureSelector(rr, 
                                n_features_to_select=40, 
                                direction="backward",
                                cv=split,
                                n_jobs=1
                               )

In [12]:
removed_columns = ["season", "date", "won", "target", "team", "team_opp"]
selected_columns = df.columns[~df.columns.isin(removed_columns)]

object_columns = df[selected_columns].select_dtypes(include='object').columns

object_columns = df[selected_columns].select_dtypes(include='object').columns

# Convert these object type columns to integers
## 1	2	3	4	total	1_opp	2_opp	3_opp	4_opp	total_opp columns
for column in object_columns:
    df[column] = df[column].astype(int)
    


In [13]:
df[selected_columns].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17744 entries, 0 to 17743
Columns: 143 entries, mp to home_opp
dtypes: float64(131), int32(10), int64(2)
memory usage: 18.7 MB


In [14]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df[selected_columns] = scaler.fit_transform(df[selected_columns])

In [15]:
from datetime import datetime

now = datetime.now()
dt_string_S = now.strftime("%d/%m/%Y %H:%M:%S")

print("Started at = ", dt_string_S)


sfs.fit(df[selected_columns], df["target"])

Started at =  27/04/2024 08:06:36


SequentialFeatureSelector(cv=TimeSeriesSplit(gap=0, max_train_size=None, n_splits=3, test_size=None),
                          direction='backward',
                          estimator=RidgeClassifier(alpha=1),
                          n_features_to_select=40, n_jobs=1)

In [16]:
predictors = list(selected_columns[sfs.get_support()])
predictors

['pf_total',
 'ftr_total',
 '3pa_max',
 '3p%_max',
 'fta_max',
 'ft%_max',
 'drb_max',
 'tov_max',
 'pts_max',
 '+/-_max',
 'efg%_max',
 'ftr_max',
 'drb%_max',
 'ast%_max',
 'fga_total_opp',
 '3p_total_opp',
 '3p%_total_opp',
 'ft_total_opp',
 'trb_total_opp',
 'ast_total_opp',
 'blk_total_opp',
 '3par_total_opp',
 'ftr_total_opp',
 'orb%_total_opp',
 'tov%_total_opp',
 'fg_max_opp',
 'fga_max_opp',
 'fg%_max_opp',
 '3p_max_opp',
 '3pa_max_opp',
 'ft_max_opp',
 'ft%_max_opp',
 'trb_max_opp',
 'tov_max_opp',
 'pf_max_opp',
 'ftr_max_opp',
 'ortg_max_opp',
 '1_opp',
 '3_opp',
 '4_opp']

1. backtest function helps to split the data according to seasons, to make sure we use atleast last 2 seasons of games to predict the game.

In [17]:
def backtest(data, model, predictors, start=2, step=1):
    all_predictions = []
    
    seasons = sorted(data["season"].unique())
    
    for i in range(start, len(seasons), step):
        season = seasons[i]
        train = data[data["season"] < season] ## Previous season
        test = data[data["season"] == season] ## Current season
        
        model.fit(train[predictors], train["target"])
        
        preds = model.predict(test[predictors])
        preds = pd.Series(preds, index=test.index)
        combined = pd.concat([test["target"], preds], axis=1)
        combined.columns = ["actual", "prediction"]
        
        all_predictions.append(combined)
    return pd.concat(all_predictions)

In [18]:
predictions = backtest(df, rr, predictors)
predictions

actual  prediction
3706        1           1
3707        0           0
3708        1           0
3709        0           1
3710        1           0
...       ...         ...
17739       0           1
17740       0           0
17741       1           1
17742       2           1
17743       2           0

[14038 rows x 2 columns]

In [19]:
from sklearn.metrics import accuracy_score

## removing unpredicted matches
predictions = predictions[predictions['actual'] != 2]
print("Accuracy Score: ",accuracy_score(predictions["actual"], predictions["prediction"]))

now = datetime.now()
dt_string_E = now.strftime("%d/%m/%Y %H:%M:%S")

print("Ended at = ", dt_string_E)


Accuracy Score:  0.5167761279268989
Ended at =  27/04/2024 08:23:12


Checking home ground emotion to check with our accuracy

In [20]:
df.groupby(["home"]).apply(lambda x: x[x["won"] == 1].shape[0] / x.shape[0])

home
0.0    0.481177
1.0    0.518823
dtype: float64

In [21]:
df.groupby(["team", "season"])['date'].count()

team                season
Atlanta Hawks       2016       33
                    2017       91
                    2018       81
                    2019       80
                    2020       37
                             ... 
Washington Wizards  2019       77
                    2020       45
                    2021      108
                    2022       84
                    2023       44
Name: date, Length: 240, dtype: int64

1. Here instead of predicting just using previous match we are trying to do it with last 10 matches
2. team.rolling(10).mean() -> Finding avg of last 10 matches

In [22]:
df_rolling = df[list(selected_columns) + ["won", "team", "season"]]

def find_team_averages(team):
    rolling = team.rolling(10).mean()
    return rolling

df_rolling = df_rolling.groupby(["team", "season"], group_keys=False).apply(find_team_averages)

df_rolling

C:\Users\JETENDRA\AppData\Local\Temp\ipykernel_50684\1305469341.py:4: FutureWarning: Dropping of nuisance columns in rolling operations is deprecated; in a future version this will raise TypeError. Select only valid columns before calling the operation. Dropped columns were Index(['team'], dtype='object')
  rolling = team.rolling(10).mean()


mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
0      NaN       NaN        NaN        NaN       NaN        NaN        NaN   
1      NaN       NaN        NaN        NaN       NaN        NaN        NaN   
2      NaN       NaN        NaN        NaN       NaN        NaN        NaN   
3      NaN       NaN        NaN        NaN       NaN        NaN        NaN   
4      NaN       NaN        NaN        NaN       NaN        NaN        NaN   
...    ...       ...        ...        ...       ...        ...        ...   
17739  0.0  0.500000   0.347761   0.576794  0.357143   0.346032   0.407605   
17740  0.0  0.393182   0.361194   0.439952  0.400000   0.377778   0.433460   
17741  0.0  0.475000   0.332836   0.560526  0.382143   0.355556   0.430672   
17742  0.0  0.465909   0.331343   0.550478  0.353571   0.357143   0.392395   
17743  0.0  0.372727   0.367164   0.412919  0.400000   0.392063   0.422180   

       ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  \
0           NaN        NaN        NaN        NaN        NaN        NaN   
1           NaN        NaN        NaN        NaN        NaN        NaN   
2           NaN        NaN        NaN        NaN        NaN        NaN   
3           NaN        NaN        NaN        NaN        NaN        NaN   
4           NaN        NaN        NaN        NaN        NaN        NaN   
...         ...        ...        ...        ...        ...        ...   
17739  0.418605   0.352381   0.785998   0.334483   0.413636   0.429412   
17740  0.316279   0.260317   0.829988   0.327586   0.313636   0.339216   
17741  0.416279   0.349206   0.788681   0.317241   0.429545   0.433333   
17742  0.397674   0.341270   0.767095   0.320690   0.443182   0.447059   
17743  0.306977   0.255556   0.821587   0.327586   0.311364   0.337255   

       ast_total  stl_total  blk_total  tov_total  pf_total  pts_total  \
0            NaN        NaN        NaN        NaN       NaN        NaN   
1            NaN        NaN        NaN        NaN       NaN        NaN   
2            NaN        NaN        NaN        NaN       NaN        NaN   
3            NaN        NaN        NaN        NaN       NaN        NaN   
4            NaN        NaN        NaN        NaN       NaN        NaN   
...          ...        ...        ...        ...       ...        ...   
17739   0.476744   0.268182      0.205   0.320690  0.341667   0.464286   
17740   0.374419   0.263636      0.165   0.324138  0.336111   0.351786   
17741   0.472093   0.290909      0.230   0.313793  0.338889   0.450000   
17742   0.446512   0.290909      0.235   0.327586  0.327778   0.428571   
17743   0.362791   0.263636      0.175   0.317241  0.347222   0.332143   

       ts%_total  efg%_total  3par_total  ftr_total  orb%_total  drb%_total  \
0            NaN         NaN         NaN        NaN         NaN         NaN   
1            NaN         NaN         NaN        NaN         NaN         NaN   
2            NaN         NaN         NaN        NaN         NaN         NaN   
3            NaN         NaN         NaN        NaN         NaN         NaN   
4            NaN         NaN         NaN        NaN         NaN         NaN   
...          ...         ...         ...        ...         ...         ...   
17739   0.575328    0.554656    0.414333   0.340360    0.452330    0.662366   
17740   0.458734    0.452834    0.450833   0.251285    0.396237    0.555197   
17741   0.572271    0.550810    0.432167   0.341388    0.427419    0.686918   
17742   0.548908    0.532794    0.434667   0.334319    0.431541    0.682258   
17743   0.432314    0.429757    0.464500   0.246144    0.381541    0.537276   

       trb%_total  ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  \
0             NaN         NaN         NaN         NaN         NaN         NaN   
1             NaN         NaN         NaN         NaN         NaN         NaN   
2             NaN         NaN         NaN         NaN         NaN         NaN   
3             NaN   

In [23]:
rolling_cols = [f"{col}_10" for col in df_rolling.columns]
df_rolling.columns = rolling_cols
df = pd.concat([df, df_rolling], axis=1)

df = df.dropna()

print("df shape: ",df.shape)
df

df shape:  (15584, 294)


mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
235    0.00  0.295455   0.358209   0.313397  0.250000   0.222222   0.415716   
242    0.00  0.272727   0.298507   0.330144  0.214286   0.333333   0.249683   
244    0.25  0.500000   0.402985   0.526316  0.178571   0.222222   0.295311   
249    0.25  0.545455   0.373134   0.607656  0.428571   0.412698   0.432193   
260    0.00  0.409091   0.432836   0.392344  0.178571   0.206349   0.313054   
...     ...       ...        ...        ...       ...        ...        ...   
17739  0.00  0.454545   0.283582   0.583732  0.142857   0.174603   0.285171   
17740  0.00  0.318182   0.253731   0.430622  0.250000   0.285714   0.338403   
17741  0.00  0.409091   0.268657   0.538278  0.464286   0.333333   0.566540   
17742  0.00  0.386364   0.343284   0.437799  0.142857   0.333333   0.159696   
17743  0.00  0.272727   0.522388   0.179426  0.285714   0.444444   0.258555   

       ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  \
235    0.325581   0.301587   0.708285   0.379310   0.363636   0.411765   
242    0.302326   0.238095   0.854142   0.413793   0.545455   0.588235   
244    0.627907   0.539683   0.766628   0.448276   0.431818   0.509804   
249    0.418605   0.412698   0.654609   0.241379   0.295455   0.274510   
260    0.325581   0.285714   0.753792   0.413793   0.318182   0.392157   
...         ...        ...        ...        ...        ...        ...   
17739  0.488372   0.412698   0.784131   0.448276   0.659091   0.705882   
17740  0.372093   0.301587   0.824971   0.275862   0.295455   0.294118   
17741  0.348837   0.317460   0.722287   0.172414   0.295455   0.235294   
17742  0.279070   0.349206   0.492415   0.379310   0.681818   0.686275   
17743  0.302326   0.238095   0.854142   0.379310   0.386364   0.431373   

       ast_total  stl_total  blk_total  tov_total  pf_total  pts_total  \
235     0.255814   0.590909       0.15   0.482759  0.361111   0.241071   
242     0.279070   0.409091       0.35   0.482759  0.333333   0.205357   
244     0.255814   0.318182       0.15   0.482759  0.416667   0.500000   
249     0.511628   0.363636       0.20   0.344828  0.500000   0.517857   
260     0.302326   0.181818       0.25   0.310345  0.111111   0.312500   
...          ...        ...        ...        ...       ...        ...   
17739   0.488372   0.136364       0.25   0.413793  0.333333   0.401786   
17740   0.372093   0.090909       0.15   0.448276  0.361111   0.276786   
17741   0.441860   0.500000       0.35   0.172414  0.333333   0.392857   
17742   0.325581   0.272727       0.35   0.448276  0.194444   0.267857   
17743   0.255814   0.409091       0.35   0.241379  0.416667   0.223214   

       ts%_total  efg%_total  3par_total  ftr_total  orb%_total  drb%_total  \
235     0.294760    0.295547    0.258333   0.286632    0.401434    0.801075   
242     0.314410    0.303644    0.423333   0.239075    0.456989    0.700717   
244     0.502183    0.449393    0.245000   0.496144    0.555556    0.700717   
249     0.596070    0.603239    0.486667   0.388175    0.299283    0.444444   
260     0.334061    0.336032    0.216667   0.255784    0.439068    0.551971   
...          ...         ...         ...        ...         ...         ...   
17739   0.530568    0.491903    0.221667   0.419023    0.646953    0.673835   
17740   0.430131    0.402834    0.381667   0.313625    0.387097    0.736559   
17741   0.572052    0.568826    0.436667   0.326478    0.263441    0.612903   
17742   0.325328    0.366397    0.401667   0.336761    0.448029    0.654122   
17743   0.179039    0.182186    0.455000   0.199229    0.345878    0.551971   

       trb%_total  ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  \
235      0.516216    0.453505    0.621495    0.158416    0.477612         0.0   
242      0.627027    0.520744    0.471963    0.299505    0.507463         0.0   
244      0.691892    0.296137    0.322430    0.141089    0.436567         0.0   
249      

Giving the prior information to modelling which we will be knowing before the start of the match

In [24]:
def shift_col(team, col_name):
    next_col = team[col_name].shift(-1)
    return next_col

def add_col(df, col_name):
    return df.groupby("team", group_keys=False).apply(lambda x: shift_col(x, col_name))

df["home_next"] = add_col(df, "home")
df["team_opp_next"] = add_col(df, "team_opp")
df["date_next"] = add_col(df, "date")
df

mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
235    0.00  0.295455   0.358209   0.313397  0.250000   0.222222   0.415716   
242    0.00  0.272727   0.298507   0.330144  0.214286   0.333333   0.249683   
244    0.25  0.500000   0.402985   0.526316  0.178571   0.222222   0.295311   
249    0.25  0.545455   0.373134   0.607656  0.428571   0.412698   0.432193   
260    0.00  0.409091   0.432836   0.392344  0.178571   0.206349   0.313054   
...     ...       ...        ...        ...       ...        ...        ...   
17739  0.00  0.454545   0.283582   0.583732  0.142857   0.174603   0.285171   
17740  0.00  0.318182   0.253731   0.430622  0.250000   0.285714   0.338403   
17741  0.00  0.409091   0.268657   0.538278  0.464286   0.333333   0.566540   
17742  0.00  0.386364   0.343284   0.437799  0.142857   0.333333   0.159696   
17743  0.00  0.272727   0.522388   0.179426  0.285714   0.444444   0.258555   

       ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  \
235    0.325581   0.301587   0.708285   0.379310   0.363636   0.411765   
242    0.302326   0.238095   0.854142   0.413793   0.545455   0.588235   
244    0.627907   0.539683   0.766628   0.448276   0.431818   0.509804   
249    0.418605   0.412698   0.654609   0.241379   0.295455   0.274510   
260    0.325581   0.285714   0.753792   0.413793   0.318182   0.392157   
...         ...        ...        ...        ...        ...        ...   
17739  0.488372   0.412698   0.784131   0.448276   0.659091   0.705882   
17740  0.372093   0.301587   0.824971   0.275862   0.295455   0.294118   
17741  0.348837   0.317460   0.722287   0.172414   0.295455   0.235294   
17742  0.279070   0.349206   0.492415   0.379310   0.681818   0.686275   
17743  0.302326   0.238095   0.854142   0.379310   0.386364   0.431373   

       ast_total  stl_total  blk_total  tov_total  pf_total  pts_total  \
235     0.255814   0.590909       0.15   0.482759  0.361111   0.241071   
242     0.279070   0.409091       0.35   0.482759  0.333333   0.205357   
244     0.255814   0.318182       0.15   0.482759  0.416667   0.500000   
249     0.511628   0.363636       0.20   0.344828  0.500000   0.517857   
260     0.302326   0.181818       0.25   0.310345  0.111111   0.312500   
...          ...        ...        ...        ...       ...        ...   
17739   0.488372   0.136364       0.25   0.413793  0.333333   0.401786   
17740   0.372093   0.090909       0.15   0.448276  0.361111   0.276786   
17741   0.441860   0.500000       0.35   0.172414  0.333333   0.392857   
17742   0.325581   0.272727       0.35   0.448276  0.194444   0.267857   
17743   0.255814   0.409091       0.35   0.241379  0.416667   0.223214   

       ts%_total  efg%_total  3par_total  ftr_total  orb%_total  drb%_total  \
235     0.294760    0.295547    0.258333   0.286632    0.401434    0.801075   
242     0.314410    0.303644    0.423333   0.239075    0.456989    0.700717   
244     0.502183    0.449393    0.245000   0.496144    0.555556    0.700717   
249     0.596070    0.603239    0.486667   0.388175    0.299283    0.444444   
260     0.334061    0.336032    0.216667   0.255784    0.439068    0.551971   
...          ...         ...         ...        ...         ...         ...   
17739   0.530568    0.491903    0.221667   0.419023    0.646953    0.673835   
17740   0.430131    0.402834    0.381667   0.313625    0.387097    0.736559   
17741   0.572052    0.568826    0.436667   0.326478    0.263441    0.612903   
17742   0.325328    0.366397    0.401667   0.336761    0.448029    0.654122   
17743   0.179039    0.182186    0.455000   0.199229    0.345878    0.551971   

       trb%_total  ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  \
235      0.516216    0.453505    0.621495    0.158416    0.477612         0.0   
242      0.627027    0.520744    0.471963    0.299505    0.507463         0.0   
244      0.691892    0.296137    0.322430    0.141089    0.436567         0.0   
249      

In [25]:
full = df.merge(df[rolling_cols + ["team_opp_next", "date_next", "team"]], left_on=["team", "date_next"], right_on=["team_opp_next", "date_next"])
full

mp  fg_total  fga_total  fg%_total  3p_total  3pa_total  3p%_total  \
0      0.25  0.500000   0.402985   0.526316  0.178571   0.222222   0.295311   
1      0.00  0.568182   0.432836   0.578947  0.107143   0.174603   0.214195   
2      0.00  0.340909   0.373134   0.358852  0.178571   0.317460   0.214195   
3      0.00  0.409091   0.537313   0.318182  0.142857   0.126984   0.354880   
4      0.00  0.590909   0.462687   0.578947  0.214286   0.285714   0.287706   
...     ...       ...        ...        ...       ...        ...        ...   
15443  0.00  0.409091   0.208955   0.600478  0.357143   0.333333   0.430925   
15444  0.00  0.295455   0.462687   0.241627  0.357143   0.444444   0.330798   
15445  0.00  0.454545   0.283582   0.583732  0.142857   0.174603   0.285171   
15446  0.00  0.318182   0.253731   0.430622  0.250000   0.285714   0.338403   
15447  0.00  0.409091   0.268657   0.538278  0.464286   0.333333   0.566540   

       ft_total  fta_total  ft%_total  orb_total  drb_total  trb_total  \
0      0.627907   0.539683   0.766628   0.448276   0.431818   0.509804   
1      0.209302   0.206349   0.666278   0.379310   0.227273   0.294118   
2      0.465116   0.428571   0.708285   0.379310   0.295455   0.352941   
3      0.511628   0.380952   0.906651   0.517241   0.477273   0.588235   
4      0.348837   0.396825   0.550758   0.413793   0.318182   0.392157   
...         ...        ...        ...        ...        ...        ...   
15443  0.418605   0.333333   0.841307   0.310345   0.295455   0.313725   
15444  0.325581   0.285714   0.753792   0.344828   0.159091   0.215686   
15445  0.488372   0.412698   0.784131   0.448276   0.659091   0.705882   
15446  0.372093   0.301587   0.824971   0.275862   0.295455   0.294118   
15447  0.348837   0.317460   0.722287   0.172414   0.295455   0.235294   

       ast_total  stl_total  blk_total  tov_total  pf_total  pts_total  \
0       0.255814   0.318182       0.15   0.482759  0.416667   0.500000   
1       0.372093   0.500000       0.05   0.275862  0.277778   0.375000   
2       0.232558   0.409091       0.45   0.413793  0.444444   0.312500   
3       0.325581   0.318182       0.30   0.310345  0.388889   0.375000   
4       0.348837   0.181818       0.30   0.310345  0.361111   0.473214   
...          ...        ...        ...        ...       ...        ...   
15443   0.372093   0.318182       0.10   0.413793  0.416667   0.392857   
15444   0.302326   0.318182       0.15   0.103448  0.444444   0.267857   
15445   0.488372   0.136364       0.25   0.413793  0.333333   0.401786   
15446   0.372093   0.090909       0.15   0.448276  0.361111   0.276786   
15447   0.441860   0.500000       0.35   0.172414  0.333333   0.392857   

       ts%_total  efg%_total  3par_total  ftr_total  orb%_total  drb%_total  \
0       0.502183    0.449393    0.245000   0.496144    0.555556    0.700717   
1       0.438865    0.469636    0.180000   0.185090    0.505376    0.372760   
2       0.334061    0.307692    0.370000   0.403599    0.458781    0.388889   
3       0.307860    0.257085    0.105000   0.316195    0.526882    0.534050   
4       0.471616    0.502024    0.300000   0.348329    0.512545    0.621864   
...          ...         ...         ...        ...         ...         ...   
15443   0.628821    0.591093    0.468333   0.361183    0.503584    0.612903   
15444   0.257642    0.259109    0.480000   0.250643    0.326165    0.353047   
15445   0.530568    0.491903    0.221667   0.419023    0.646953    0.673835   
15446   0.430131    0.402834    0.381667   0.313625    0.387097    0.736559   
15447   0.572052    0.568826    0.436667   0.326478    0.263441    0.612903   

       trb%_total  ast%_total  stl%_total  blk%_total  tov%_total  usg%_total  \
0        0.691892    0.296137    0.322430    0.141089    0.436567         0.0   
1        0.413514    0.412017    0.570093    0.042079    0.283582         0.0   
2        0.391892    0.371960    0.453271    0.405941    0.399254         0.0   
3        

In [26]:
full[["team_x", "team_opp_next_x", "team_y", "team_opp_next_y", "date_next"]]

team_x         team_opp_next_x                  team_y  \
0           Sacramento Kings       San Antonio Spurs       San Antonio Spurs   
1            Detroit Pistons   Oklahoma City Thunder   Oklahoma City Thunder   
2             Indiana Pacers           Orlando Magic           Orlando Magic   
3              Chicago Bulls  Portland Trail Blazers  Portland Trail Blazers   
4      Oklahoma City Thunder         Detroit Pistons         Detroit Pistons   
...                      ...                     ...                     ...   
15443         Denver Nuggets              Miami Heat              Miami Heat   
15444             Miami Heat          Denver Nuggets          Denver Nuggets   
15445         Denver Nuggets              Miami Heat              Miami Heat   
15446             Miami Heat          Denver Nuggets          Denver Nuggets   
15447         Denver Nuggets              Miami Heat              Miami Heat   

             team_opp_next_y   date_next  
0           Sacramento Kings  2016-11-16  
1            Detroit Pistons  2016-11-14  
2             Indiana Pacers  2016-11-14  
3              Chicago Bulls  2016-11-15  
4      Oklahoma City Thunder  2016-11-14  
...                      ...         ...  
15443         Denver Nuggets  2023-06-07  
15444             Miami Heat  2023-06-09  
15445         Denver Nuggets  2023-06-09  
15446             Miami Heat  2023-06-12  
15447         Denver Nuggets  2023-06-12  

[15448 rows x 5 columns]

In [27]:
removed_columns = list(full.columns[full.dtypes == "object"]) + removed_columns
removed_columns

['team_x',
 'team_opp',
 'date',
 'team_opp_next_x',
 'date_next',
 'team_opp_next_y',
 'team_y',
 'season',
 'date',
 'won',
 'target',
 'team',
 'team_opp']

In [ ]:
selected_columns = full.columns[~full.columns.isin(removed_columns)]


from datetime import datetime

now = datetime.now()
dt_string_S = now.strftime("%d/%m/%Y %H:%M:%S")

print("Started at = ", dt_string_S)

sfs.fit(full[selected_columns], full["target"])

Started at =  27/04/2024 08:23:15


In [ ]:
predictors = list(selected_columns[sfs.get_support()])
predictors

In [ ]:
predictions = backtest(full, rr, predictors)

## removing unpredicted matches
predictions = predictions[predictions['actual'] != 2]
predictions

In [ ]:


print("Accuracy Score: ",accuracy_score(predictions["actual"], predictions["prediction"]) * 100)

now = datetime.now()
dt_string_E = now.strftime("%d/%m/%Y %H:%M:%S")

print("Ended at = ", dt_string_E)


1. last 2 seasons backtest + 10 rolling (30 Features forward) accuracy: 54 % 
2. last 2 seasons backtest + 15 rolling (30 Features forward) accuracy: 53.7 % 
3. last 2 seasons backtest + 5 rolling (30 Features forward) accuracy: 52.8 %

Improving model:
1. try powerful model
2. no of features & try backward method
3. Try rolling mean (with different combinations)
4. Make sure no null matches (before full dataframe or get only till last season and try)